File: backend/app/schemas/signals.py
from typing import Literal

from pydantic import BaseModel, Field


SignalKind = Literal[
    "whale-flow",
    "order-flow",
    "volume-anomaly",
    "price-move",
    "poll-divergence",
    "normal",
]
SignalSeverity = Literal["normal", "elevated", "high", "critical"]
SignalSource = Literal["fixture", "live"]
SignalBatchSource = Literal["fixture", "mixed", "live"]


class SignalComponentResponse(BaseModel):
    key: str
    label: str
    value: float = Field(ge=0, le=100)
    weight: float = Field(ge=0, le=1)
    contribution: float = Field(ge=0)
    available: bool
    detail: str


class RegionSignalResponse(BaseModel):
    region_code: str = Field(alias="regionCode")
    country_code: str = Field(alias="countryCode")
    market_slug: str = Field(alias="marketSlug")
    kind: SignalKind
    score: int = Field(ge=0, le=100)
    severity: SignalSeverity
    headline: str
    detail: str
    observed_at: str = Field(alias="observedAt")
    source: SignalSource
    confidence: float = Field(ge=0, le=1)
    baseline_window: str = Field(alias="baselineWindow")
    components: list[SignalComponentResponse]

    model_config = {"populate_by_name": True}


class RegionSignalsResponse(BaseModel):
    country_code: str = Field(alias="countryCode")
    generated_at: str = Field(alias="generatedAt")
    source: SignalBatchSource
    signals: list[RegionSignalResponse]

    model_config = {"populate_by_name": True}

File: src/types/signals.ts
export type MarketSignalKind =
  | "whale-flow"
  | "order-flow"
  | "volume-anomaly"
  | "price-move"
  | "poll-divergence"
  | "normal";

export type MarketSignalSource = "fixture" | "live";

export type MarketSignalSeverity = "inactive" | "normal" | "elevated" | "high" | "critical";

export type RegionSignalComponent = {
  key: string;
  label: string;
  value: number;
  weight: number;
  contribution: number;
  available: boolean;
  detail: string;
};

export type RegionMarketSignal = {
  kind: MarketSignalKind;
  score: number;
  headline: string;
  detail: string;
  observedAt: string;
  source: MarketSignalSource;
  confidence?: number;
  baselineWindow?: string;
  components?: RegionSignalComponent[];
};

export type RegionSignal = RegionMarketSignal & {
  regionCode: string;
  countryCode: string;
  marketSlug: string;
  severity: Exclude<MarketSignalSeverity, "inactive">;
  confidence: number;
  baselineWindow: string;
  components: RegionSignalComponent[];
};

export type RegionSignalsResponse = {
  countryCode: string;
  generatedAt: string;
  source: "fixture" | "mixed" | "live";
  signals: RegionSignal[];
};
